Port ATen/native to ATen/Parallel

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #20157 Support for Eigen thread pool
* #20087 Native ATen/Parallel backend
* #20057 Split ATen/Parallel into interface and backend
* #20050 Move inter-op settings into ATen/Parallel
* **#20043 Port ATen/native to ATen/Parallel**
* #20032 Port THNN to ATen/Parallel

Summary:
Remove explicit usage of OpenMP from ATen/native

Test Plan:
BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

Differential Revision: [D15248505](https://our.internmc.facebook.com/intern/diff/D15248505)

Commit:
Port ATen/native to ATen/Parallel

Summary:
Remove explicit usage of OpenMP from ATen/native

Test Plan:
BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

Commit:
Update on "Port ATen/native to ATen/Parallel"

Port ATen/native to ATen/Parallel

Summary:
Remove explicit usage of OpenMP from ATen/native

Test Plan:
BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20043 gh/ilia-cher/12/head

dzhulgakov commented on aten/src/ATen/native/Activation.cpp:
```diff
@@ -114,26 +115,31 @@ void inline prelu_cpu_kernel_multi_weights(
-      n_result_data += input_stride1;
     }
+  };
+  if (input.numel() > 1000) {
```
so your approach for now is to just preserve the original conditions? Do you plan to do another pass to figure out how to describe "minimal amount of work" to at::parallel?

ilia-cher replied:
yes, I preserve what we use. I'm pretty sure that those constants were very approximate and ad-hoc and precise tuning is needed, it would require a separate effort.

dzhulgakov commented on aten/src/ATen/native/Activation.cpp:
```diff
@@ -114,26 +115,31 @@ void inline prelu_cpu_kernel_multi_weights(
-        scalar_t w = (n_input_data[k] > 0) ? scalar_t(1) : weight_data[j];
-        n_result_data[k] = w * n_input_data[k];
+  auto loop = [&](int64_t start, int64_t end) {
+    for (auto i = 0; i < end; ++i) {
```
tests are failing probably because of `0` -> `start` here. I'm worried we might have missed other places too. If you compile it with unused variable warnings - does it produce more warnings?

ilia-cher replied:
oops, missed one, thanks - I am definitely going to audit all the places once more time

dzhulgakov commented on aten/src/ATen/native/Activation.cpp:
```diff
@@ -196,26 +202,28 @@ void inline prelu_cpu_backward_kernel_share_weights(
+  scalar_t sum = at::parallel_reduce(0, input_numel, 1000, scalar_t(0),
+      [&](int64_t start, int64_t end, scalar_t ident) -> scalar_t {
+    scalar_t partial_sum = ident;
+    for (auto i = 0; i < input_numel; i++) {
```
also no usages of start/end?

ilia-cher replied:
thanks, will do another pass

dzhulgakov reviewed (changes requested):
heroic work!

Commit:
Update on "Port ATen/native to ATen/Parallel"

Port ATen/native to ATen/Parallel

Summary:
Remove explicit usage of OpenMP from ATen/native

Test Plan:
BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20043 gh/ilia-cher/12/head

ilia-cher commented:
bugfix + double checked all other call sites

ilia-cher commented on aten/src/ATen/native/Embedding.cpp:
```diff
@@ -105,44 +102,29 @@ Tensor embedding_dense_backward_cpu(
 
-#ifdef _OPENMP
+  auto parallel_section = [&](int64_t start, int64_t end) {
+    for (int64_t i = 0; i < numel; i++) {
```
btw, note that here this is correct, we parallelize inside the loop

dzhulgakov reviewed (approved):
ok, I'll trust your audit :)

Commit:
Update on "Port ATen/native to ATen/Parallel"

Port ATen/native to ATen/Parallel

Summary:
Remove explicit usage of OpenMP from ATen/native

Test Plan:
BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20043 gh/ilia-cher/12/head

Commit:
Update on "Port ATen/native to ATen/Parallel"

Port ATen/native to ATen/Parallel

Summary:
Remove explicit usage of OpenMP from ATen/native

Test Plan:
BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20043 gh/ilia-cher/12/head

Commit:
Update on "Port ATen/native to ATen/Parallel"

Port ATen/native to ATen/Parallel

Summary:
Remove explicit usage of OpenMP from ATen/native

Test Plan:
BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20043 gh/ilia-cher/12/head

Commit:
Update on "Port ATen/native to ATen/Parallel"

Port ATen/native to ATen/Parallel

Summary:
Remove explicit usage of OpenMP from ATen/native

Test Plan:
BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20043 gh/ilia-cher/12/head

Commit:
Update on "Port ATen/native to ATen/Parallel"

Port ATen/native to ATen/Parallel

Summary:
Remove explicit usage of OpenMP from ATen/native

Test Plan:
BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20043 gh/ilia-cher/12/head

Commit:
Update on "Port ATen/native to ATen/Parallel"

Port ATen/native to ATen/Parallel

Summary:
Remove explicit usage of OpenMP from ATen/native

Test Plan:
BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20043 gh/ilia-cher/12/head

Commit:
Update on "Port ATen/native to ATen/Parallel"

Port ATen/native to ATen/Parallel

Summary:
Remove explicit usage of OpenMP from ATen/native

Test Plan:
BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20043 gh/ilia-cher/12/head

Commit:
Update on "Port ATen/native to ATen/Parallel"

Port ATen/native to ATen/Parallel

Summary:
Remove explicit usage of OpenMP from ATen/native

Test Plan:
BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20043 gh/ilia-cher/12/head

Commit:
Update on "Port ATen/native to ATen/Parallel"

Port ATen/native to ATen/Parallel

Summary:
Remove explicit usage of OpenMP from ATen/native

Test Plan:
BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20043 gh/ilia-cher/12/head

Commit:
Update on "Port ATen/native to ATen/Parallel"

Port ATen/native to ATen/Parallel

Summary:
Remove explicit usage of OpenMP from ATen/native

Test Plan:
BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20043 gh/ilia-cher/12/head

Commit:
Update on "Port ATen/native to ATen/Parallel"

Port ATen/native to ATen/Parallel

Summary:
Remove explicit usage of OpenMP from ATen/native

Test Plan:
BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
pytest -s -v test/test_torch.py::TestTorch

gh-metadata: pytorch pytorch 20043 gh/ilia-cher/12/head
